mousestyles/ultradian/__init__.py

```diff
@@ -0,0 +1,358 @@
+            end_time = end[i]
+            start_index = int(start_time / (bin_width * 60))
+            end_index = int(end_time / (bin_width * 60))
+            if start_index == end_index:
+                time_movements[start_index + j *
+                               bin_count] += dist[i]
+            else:
+                time_movements[
+                    end_index + j * bin_count] += end_time % \
+                    bin_length / (end_time - start_time) * dist[i]
+                time_movements[
+                    start_index + j * bin_count] += dist[i] - \
+                    end_time % bin_length / (end_time - start_time) * dist[i]
+    ts = pd.Series(time_movements, index=pd.date_range(
+        '01/01/2014', periods=len(time_movements),
+        freq=str(bin_width) + 'min'))
+    return(ts)
+
+
+def aggregate_data(feature, bin_width):
+    r"""
+    Aggregate all the strains and mouses with any feature together
+    in one dataframe. It combines the results you got from
+    aggregate_movements and aggregate_interval. It will return
+    a dataframe with three variables: mouse, strain, feature and hour.
+
+    Parameters
+    ----------
+    feature :
+        {"AS", "F", "IS", "M_AS", "M_IS", "W"}
+    bin_width : int
+        Number of minutes, the time interval for data aggregation.
+
+    Returns
+    -------
+    pandas.dataframe
+    describe :
+        Column 0: the mouse number (number depends on strain)(0-3)
+        Column 1: the strain of the mouse (0-2)
+        Column 2: hour(numeric values below 24 accourding to bin_width)
+        Column 3: feature values
+
+    Examples
+    --------
+    >>> test = aggregate_data("Distance",20)
+    >>> print(np.mean(test["Distance"]))
+    531.4500177747973
+    """
+    init = pd.DataFrame(columns=["mouse", "strain", "hour", feature])
+    for i in range(3):
+        for j in range(4):
+            if feature == "Distance":
+                tmp = aggregate_movement(strain=i, mouse=j,
+                                         bin_width=bin_width)
+            else:
+                tmp = aggregate_interval(strain=i, mouse=j,
+                                         feature=feature,
+                                         bin_width=bin_width)
+            tmp = pd.DataFrame(list(tmp.values), index=tmp.index)
+            tmp.columns = [feature]
+            tmp["strain"] = i
+            tmp["mouse"] = j
+            tmp["hour"] = tmp.index.hour + tmp.index.minute / 60
+            init = init.append(tmp)
+    return(init)
+
+
+def seasonal_decomposition(strain, mouse, feature, bin_width, period_length):
+    """
+    Apply seasonal decomposition model on the time series
+    of specified strain, mouse, feature and bin_width.
+
+    Parameters
+    ----------
+    strain: int
+        nonnegative integer indicating the strain number
+    mouse: int
+        nonnegative integer indicating the mouse number
+    feature: {"AS", "F", "IS", "M_AS", "M_IS", "W"}
```

Comment:
I think you missed the feature `Distance` calculated from `aggregate_movement`. We're supposed to do seasonal decomposition for that time series too.@boyinggong